src/Features/Core/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.cs

```diff
@@ -137,87 +230,446 @@ public override Task<ImmutableArray<DiagnosticData>> GetSpecificCachedDiagnostic
+                    lock (_updateDiagnosticsLock)
+                    {
+                        if (!DocumentAnalyzed(documentId) && !CompletionStarted)
+                        {
+                            _analyzedDocuments[documentId] = Task.Run(async () =>
+                            {
+                                diagnostics = await GetDocumentDiagnosticsAsync(document, cancellationToken).ConfigureAwait(false);
+                                DistributeDiagnostics(diagnostics);
+                            }, cancellationToken);
+                        }
+                    }
+
+                    if (DocumentAnalyzed(documentId))
+                    {
+                        await _analyzedDocuments[documentId].ConfigureAwait(false);
+                    }
+                }
+
+                return diagnostics;
+            }
+
+            private async Task<ImmutableArray<Diagnostic>> GetDocumentDiagnosticsAsync(Document document, CancellationToken cancellationToken)
+            {
+                SyntaxTree documentSyntax = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+                SemanticModel documentModel = _compilationWithAnalyzers.Compilation.GetSemanticModel(documentSyntax);
+                return await _compilationWithAnalyzers.GetAnalyzerDiagnosticsFromDocumentAsync(documentModel).ConfigureAwait(false);
+            }
+
+            public async Task<ImmutableArray<Diagnostic>> GetAllDiagnosticsAsync()
+            {
+                if (CompletionStarted)
+                {
+                    await _completionTask.ConfigureAwait(false);
+
+                    ImmutableArray<Diagnostic>.Builder result = ImmutableArray.CreateBuilder<Diagnostic>();
+                    lock (_updateDiagnosticsLock)
+                    {
+                        foreach (ImmutableArray<Diagnostic> diagnostics in _diagnosticsPerDocuments.Values)
+                        {
+                            result.AddRange(diagnostics);
+                        }
+                    }
+
+                    return result.ToImmutable();
                 }
 
-                Owner.RaiseDiagnosticsUpdated(
-                    this, new DiagnosticsUpdatedArgs(
-                        ValueTuple.Create(this, kv.Key), workspace, solution, project.Id, kv.Key, kv.ToImmutableArrayOrEmpty()));
+                return ImmutableArray<Diagnostic>.Empty;
             }
+
+            public async Task<ImmutableArray<Diagnostic>> GetDocumentDiagnosticsAsync(Document document)
+            {
+                if (_compilationWithAnalyzers != null)
+                {
+                    if (DocumentAnalyzed(document.Id))
+                    {
+                        await _analyzedDocuments[document.Id].ConfigureAwait(false);
+                    }
+                    else if (CompletionStarted)
+                    {
+                        await _completionTask.ConfigureAwait(false);
```

Comment:
So document diagnostics requests after project analysis has started wait on project analysis to complete?